### ISSUE TITLE
Error: "could not amend shell profile" when HOME is unwritable

### ISSUE THREAD

**Issue by jeroen** [rust-lang/rustup#2040]
**Problem**

When `CARGO_HOME` and `RUSTUP_HOME` point to a location outside of `$HOME`, the rustup installer still tries to write to `$HOME/.profile` and then errors with:

```
error: could not amend shell profile: '/Users/builder/.profile'
info: caused by: could not write rcfile file: '/Users/builder/.profile'
info: caused by: Permission denied (os error 13)
```

I can't find how to prevent rustup from touching the user `.profile` or proceed with installation even though it fails to update `.profile`.

**Steps**

We run `rustup` on a somewhat restricted and perishable MacOS 10.11 virtual machine where user `$HOME` is unwritable. Hence on each build, we need to install cargo like so:

```sh
export CARGO_HOME="${TMPDIR}/.cargo"
export RUSTUP_HOME="${TMPDIR}/.rustup"
curl https://sh.rustup.rs -sSf | bash -s -- -y
```

And then in another script, we can use the following command to build stuff:

```sh
CARGO_HOME=${TMPDIR}/.cargo RUSTUP_HOME=${TMPDIR}/.rustup  \
  ${CARGO_HOME}/bin/cargo build --release --manifest-path=myrustlib/Cargo.toml
```

This works, but the rustup installation error about `$HOME/.profile` keeps showing.

**Possible Solutions**

 1. A variable or option to opt-out of automatically updating the user `.profile`.
 2. Turn this into a warning instead of an error, such that installation can proceed.

**Notes**

- System: MacOS 10.11
- Rustup: latest

**Comment by kinnison** [comment]
You probably want to pass `--no-modify-path` to the installer.

**Comment by jeroen** [comment]
Thanks. Where do I pass this exactly? Like so ?

```
curl https://sh.rustup.rs -sSf | bash -s -- -y --no-modify-path
```

**Comment by kinnison** [comment]
Yep that's exactly right.  The `--no-modify-path` is equivalent to telling the interactive installer that you don't want it to try and alter your profile.  It will still write the `env` file to `CARGO_HOME` for you to use.

**Comment by jeroen** [comment]
Thanks!

**Comment by Gargo** [comment]
I have the same issue but currently the command is:

curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh

this path doesn't support your `--no-modify-path`

**Comment by kinnison** [comment]
add `-s -- --no-modify-path`?

**Comment by changsongl** [comment]
> I have the same issue but currently the command is:
> 
> curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
> 
> this path doesn't support your `--no-modify-path`

````shell
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | bash -s -- -y --no-modify-path
````

**Comment by pYtoner** [comment]
The error is not helpful at all, especially for beginners. Should it be changed to something better? For example, providing some solutions.

**Comment by narhzih** [comment]
During the installing, the bash_profile for your system will be modified for some reasons so what I did was that, I made the file writetable with:

 ```chmod 777 ~/.bash_profile```.
 
 Running the command after that should not throw any error. You can then return the file permission to what it was with
 `chmod 644 ~/.bash_profile`.
 Although, making the .bash_profile writable by groups and other users might have some security issues attached to it but, It does fix the issue for the period of installing what you intend to install.
 
 Also, don't forge to return the file permission back to what it was initially, it's very important you do that

**Comment by Dexter706** [comment]
Append the ~/.bash_profile with `export PATH=$PATH:/Users/$USER/.cargo/bin`  and reload it  `source ~/.bash_profile`
Now execute the cmd : `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh`

**Comment by corneliusroemer** [comment]
This issue seems to be quite common, see Stackoverflow: https://github.com/rust-lang/rustup/issues/2040

It should probably be reopened as it hinders accessibility. The script should provide a more useful error message.

**Comment by eldoy** [comment]
I got this weird error about fish shell when trying to install rust even though I don't even have fish shell installed:

```
error: could not amend shell profile: '/Users/vidar/.config/fish/conf.d/rustup.fish': could not write rcfile file: '/Users/vidar/.config/fish/conf.d/rustup.fish': No such file or directory (os error 2)
```

Resolved it by using `--no-modify-path`.

**Comment by rami3l** [comment]
> I got this weird error about fish shell when trying to install rust even though I don't even have fish shell installed:
> 
> ```
> error: could not amend shell profile: '/Users/vidar/.config/fish/conf.d/rustup.fish': could not write rcfile file: '/Users/vidar/.config/fish/conf.d/rustup.fish': No such file or directory (os error 2)
> ```
> 
> Resolved it by using `--no-modify-path`.

@eldoy FYI, this is #3706. A fix is available on master and we expect to ship v1.27.1 very soon.

I'm glad that you have found a workaround, although according to our policy we only write fish rcfiles if you have fish installed, so what you said sounds... well, a bit fishy... 🐟

**Comment by xanperna** [comment]
> You probably want to pass `--no-modify-path` to the installer.

To beginners like me, people like you are absolutely amazing. Like seriously how the hell do you guys just know how to fix stuff like this, you guys are like wizards to me.
